include/rdma/fi_profile.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2023 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_PROFILE_H
+#define FI_PROFILE_H
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <rdma/fi_domain.h>
+
+/*
+ * pre-defined profiling variables
+ */
+enum {
+	FI_VAR_UNEXP_MSG_CNT,    // type: uint64_t
+};
+
+/*
+ * pre-defined events which can change the profiling variables
+ */
+enum {
+	FI_EVENT_UNEXP_MSG_RECVD,    // var = FI_VAR_UNEXP_MSG_CNT
+	FI_EVENT_UNEXP_MSG_MATCHED,  // var = FI_VAR_UNEXP_MSG_CNT
+};
+
+struct fi_profile_desc {
+	uint32_t id;
+	enum fi_datatype type;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+
+struct fid_profile;
+
+struct fi_profile_ops {
+	size_t size;
+	void (*reset)(struct fid_profile *prof_fid,  uint64_t flags);
+	int (*query_vars)(struct fid_profile *prof_fid,
+	                     struct fi_profile_desc *varlist, int *count);
+	int (*query_events)(struct fid_profile *prof_fid,
+	                     struct fi_profile_desc *eventlist, int *count);
+	int (*read_var)(struct fid_profile *prof_fid, uint32_t var_id, void *data);
```

Comment:
Based on @swelch 's comments about pulling the unexpected list, we may want to change this function to return ssize_t instead of an int, plus add a size_t *size parameter.  We need to discuss how we want a pull model to work when the data being returned may be variable sized.  There are other options here, such as adding iterator calls.

Reply:
Will change op for pull data:
ssize_t (*read_var)(struct fid_profile *prof_fid, uint32_t var_id, void *data, size_t *count);
        on input, *count is the max entries in the data.  On output, *count is the total entries currently 
         available for the var_id.  The return is the actual count of entries in the data buffer if successed.

callback for pushing data will be:
define a new struct, 
struct event_data {
        void *data;
        size_t count;    // # of entries in the data
};
int (*callback)(struct fid_profile *prof_fid, uint32_t event, uint64_t flags, struct event_data *data,
                       size_t count, void *context);
         flags can set control bit to limit pulling data in callback; 
         data will be the array of event_data struct; and 
         the count will be the # of event data in the data array. 

In this way, we should be able to support the "ex_get_unexpected_msg()" call below in either pull or push mode.
This will allow one call to pull or push data as one snapshot.

Reply:
For read_var and the callback, I would use size instead of count.  Since data is a void *, I would allow the size to be interpreted as bytes, which provides more flexibility.  We don't know what future events may need to report.

We have the option to interpret size as a count for a given var_id or event, but I suspect that may end up confusing.

Please shift flags in the callback to the second to last parameter, just before void * context.  That aligns best with other APIs.  We don't need to define any flags for now, just having the parameter gives us flexibility in the future.

For the callback, it's easier if the provider can just return an array of whatever data they need.  The struct event_data that's defined above makes things difficult on the provider.  It forces them to allocate 2 arrays and link them together.

So...
```
ssize_t (*read_var)(prof_fid, u32 var_id, void *data, size_t *size);

int (*callback)(prof_fid, u32 event, void *data, size_t size, u64 flags, void *context);
```
To cover the pull model, we'll need a new variable, say FI_VAR_UNEXP_QUEUE.  That can use Ian's updated struct fi_cq_err_entry for output.  And that leads us to... somewhat of a mess.

enum fi_datatype works great for simple data types.  It's not so good for structs.  But enum fi_type is good for structs.  It even makes sense to add support for FI_TYPE_CQ_ERR_ENTRY.

What if we include both enum fi_type and enum fi_datatype in fi_profile_desc (maybe as a union)?  Then we just need a way to indicate which one applies, such as adding a u64 flags to the fi_profile_desc.

Reply:
Will add fi_type as union with fi_datatype.  Since we need to add a flags in fi_profile_desc, we can use a bit in there to indicate if the variable can be read inside a callback. We can remove the flags in the callback.